RFC: introduce `reactNativeManifest` to package.json (#588)
react-native-community/discussions-and-proposals: proposals/0012-introduce-reactNativeMetadata.md

---
title: Introducing `reactNativeMetadata` to `package.json`, for RN specific metadata
author:
  - Lorenzo Sciandra
date: 25-01-2021
---

> # RFC0012: Introducing `reactNativeMetadata` to `package.json`, for RN specific metadata

Reviewer: ```suggestion
# RFC0588: Introducing `reactNativeMetadata` to `package.json`, for RN specific metadata
```

I believe this was discussed in the past to use PR numbers as rfc number

## Summary

This RFC wants to introduce and formally discuss a new section for the package.json of a react-native project (app or library), called `reactNativeMetadata`.

This new section will allow developer to express certain characteristics of their code in a more formalized manner, that can easily used by tooling to act towards the code accordingly.

**Note well**: this is but the first draft of the actual shape and fields. The goal of this draft is to collect feedback and help finalize the spec for this section.

## Basic examples

This is how this section would look like...

...for a library:

```json
{
    "name": "@rnx-kit/metro-serializer",
    "version": "1.0.11",
    ...,
    "reactNativeMetadata": {
        "version": "2.3",
        "type": "library",

>         "features": {

Reviewer 2: ```suggestion
        "environments": {
```

Reviewer: Resolving as this is now `capabilities`

            "fabric": "true",
            "android": {
                "turbomodules": "true",
            },
        },

>         "requirements": {

Reviewer 2: ```suggestion
        "allowedDependencies": {
```

Reviewer: Removing as this has been removed to reduce scope

            "react-native": ">= 0.71.0",
            "expo": ">= 46.0.0",
            "expo-modules-core": "1.2.0"
        }
    }
}
```

...for an app:

```json
{
    "name": "contoso",
    "version": "2.0.3",
    ...,
    "reactNativeMetadata": {
    "version": "1.0",
    "type": "app",

>     "features": {

Reviewer 2: I'm not sure that `features` here makes as much sense as it does in the library scenario.

Reviewer: Resolving as this is now `capabilities`

>     "features": {

Reviewer 2: ```suggestion
    "build": {
```

        "newArchEnabled": "true",
        "android": {
            "codegen": "true",
            "turbomodules": "true",
        },
        // Future example

>         // "metroExportsMode": "strict"

Reviewer: nit: Can we make this a valid JSON as Github renders this in a terrible way:
<img width="994" alt="Screenshot 2023-01-27 at 18 48 49" src="https://user-images.githubusercontent.com/3001957/215158164-448f25d4-7ebc-428c-af47-34eefddace13.png">

>         // "metroExportsMode": "strict"

Reviewer 2: If this is for an app, should the metro configuration be happening in the `metro.config.js`?

Reviewer: Yes, Metro config should live inside `metro.config.js`. Also this example has been removed from the proposal, so I'm resolving this comment

    },
    ...
    }
```

## Motivation

The conversation that originated this idea was [an ask from the Meta team](https://github.com/microsoft/rnx-kit/issues/1863) to have `align-deps` help developers in the community migrating to the new architecture by providing insights into which libraries would have Fabric/TM support. Very quickly it became clear that there is currently no straightforward way to know if a library indeed has that support.

> A similar problem was also encountered by [React Native Directory](https://reactnative.directory/), that tried to [programmatically infer](https://github.com/react-native-community/directory/pull/870) support via the presence of certain fields in the package.json of a library - this because one can get access to the manifest file via npm without having to download the entire node_module.

Reviewer: ```suggestion
A similar problem was also encountered by [React Native Directory](https://reactnative.directory/), that tried to [programmatically infer](https://github.com/react-native-community/directory/pull/870) support via the presence of certain fields in the package.json of a library - this because one can get access to the manifest file via npm without having to download the entire NPM package.
```

This RFC is the result of that meeting: a new section of the `package.json`, called `reactNativeMetadata`, to contain certain metadata about a react-native project - it being either a library or an app.

There are a few areas that this section could express in a consistent manner; here's a list of the ones we could think of (in no particular order):

- if a library supports the new architecture, turbo modules or fabric or both
- which versions of react-native a library supports in a given release
- if an app has opted in into Fabric, TM, or both
- if an app is using Hermes or another js engine

Sidenote: the name `reactNativeMetadata` was chosen to mirror the camel case pattern used by other fields such as `peerDependenciesMeta`.

## Detailed design

This is the section that in this first draft will need to be properly finalised. For now, here's what we envisioned:

```js
reactNativeMetadata: {
    version: string, // ex. 1.0.0 - we know that this spec might evolve in the future
    type: string; // "app" or "library"
    features: {
        // can be global...

>         newArchEnabled: boolean,

Reviewer 3: Would this be better future-proofed by doing something similar to ```jsEngine```

```suggestion
        architecture: "new", // ex. "new", "old", "futureMagicalArch"
```

Reviewer: Resolving as I believe this is not relevant anymore with the new proposal. Please read through again and share your concerns if any 👍

        jsEngine: string, // ex. "hermes", "jsc", "V8"...
        // ...or platform specific
        android: {
            newArchEnabled: boolean,
            fabric: boolean,
            turbomodules: boolean,
            jsEngine: string,
        },
        ios: {
        ... // same as android
        },
        macos: {
        ... // same as android
        },
        windows: {
        ... // same as android
        },
        // more platforms, params - add a comment here
    },
    // this section would most likely only be relevant for libraries
    requirements: {
        "react-native": string, // semver compliant range of versions of RN supported
        expo: string, // semver compliant range of versions of Expo supported (if any)
    }
}
```

## Drawbacks

> The main drawback is that we'd need to convince developers to start using it, so we'd have to expect a phase in which only a part of the userbase has this section configured and tool developers need to account for the section missing entirely.

Reviewer: nit: Is it worth investigating a way to use the directory to "distribute" pull requests to known libraries to introduce this block (or have a community effort around it to do it manually)?

## Alternatives

> `reactNativeMetadata` is, by design, fully original and new: there are other files that are used to define configurations for various aspect of react-native based projects (such as `react-native.config.js`, `app.json`, Expo's `app.config.js`, `expo-module.config.json`) but this one does **not** overlap with any of them. Explicitly, this one has the purpose of being filled only with metadata to be read by package managers and other tools.

Reviewer: > to be read by package managers and other tools.

I think we should add here what should be the semantic if say `jsEngine: "hermes"` is defined in `package.json` but the `ios/Podfile` or `android/app/gradle.properties` file are disagreeing.

I believe `package.json` values should take higher priority above everything.

Reviewer: Clarified on 76cb7db062beb14af3a610a87bce49785f51daa2

This RFC introduces and formally discuss the concept of _React Native Manifest_, a set of metadata about a React Native project (either app or library). This metadata can be consumed by tooling to enhance the developer experience and support the React Native ecosystem.

Practically, this RFC proposes to add a `reactNativeManifest` section to the `package.json` file of a React Native app or library.
This new section will allow developers to follow a declarative approach to express capabilities of apps and libraries in a more formalized manner.

## Rationale

The need to specify **capabilities and metadata** for a react-native library is inherent to the React Native ecosystem.

As of today, there is no simple programmatic way to know if a NPM package is a React Native library. Good indicators could be:
* Presence of a `react-native:*` dependency in the `peerDependencies` section of the `package.json` ([example](https://github.com/RonRadtke/react-native-blob-util/blob/80628d418a5c81439dc2c1a1f05fae6406f0ba7f/package.json#L46C10-L49))
* Presence of React Native specific files such as `react-native.config.js` files

> Those indicators are all non-exhaustive. 

Reviewer 4: How do other libraries determine this? I guess it's not really a problem because we need to be able to catalogue libraries specifically for React Native?

Reviewer: > How do other libraries determine this?

What do you mean with "other libraries". Like other ecosystems?

Reviewer 4: Oops yes other ecosystems. I'm guessing they have similar directories. Are they all maintaining it manually?

> Similarly, there is no easy way to understand if a library is exposing a native module or not. 

Reviewer 4: What's the logic for needing to know whether a library exposes a native module or not?

Reviewer: > Why do we need to know whether a library exposes a native module or not?

For example: today we cannot crawl NPM to populate https://reactnative.directory/
Libraries needs to be added manually.

  * The alternative at this stage is to inspect the code and check the API callsites if they are New Architecture or Old Architecture compatible.
* Lack of a declarative way to enable New Architecture on both platforms
  * Currently New Architecture is enabled by a Gradle Property `newArchEnabled` on Android in the `gradle.properties` file and by invokign `RCT_NEW_ARCH_ENABLED pod install` on iOS.
  * As of today, there is no way to enable New Architecture for an app project for both platforms
  * Moreover, the file vs environment variable leads to a scenario where you can't know if a app supports New Architecture or not by inspecting the code (as New Architecture support is known at  `pod install` time and is not codified in the codebase)

Therefore we propose to add the `newArch` section to the `reactNativeManifest.capabilities` of **both apps and libraries** with the following semantic:
* For Apps: `newArch.enabled==true` implies that the app wants to use the New Architecture.

> * For Libraries: `newArch.enabled==true` implies that the library is compatible with the New Architecture.

Reviewer 4: Is the thinking then that there's no way to express one platform only supporting newArch? And it's not something we want to support?

Tools can be built on top of this information to check that a app with `newArch.enabled==true` is not accepting libraries with `newArch.enabled==false` and warning against library that don't have the key specified.

The setup would look as follows for both apps and libraries:

```json
{
  "reactNativeManifest": {
    "capabilities": {
      "newArch": {
        "enabled": true
      }
    }
  }
}
```

This section will allows also for split configuration between platforms:

```json
{
  "reactNativeManifest": {
    "android": {
      "capabilities": {
        "newArch": {
          "enabled": true
        }
      }
    },
    "ios": {
      "capabilities": {
        "newArch": {
          "enabled": true
        }
      }
    }
  }

> }

Reviewer: Nope it's not. Top level capabilities applies to all platforms ("windows", "macos", others).

* [android/app/build.gradle](https://github.com/facebook/react-native/blob/main/packages/react-native/template/android/app/build.gradle) to specify build time configuration for Android
* [ReactNativeFeatureFlags.js](https://github.com/facebook/react-native/blob/main/packages/react-native/Libraries/ReactNative/ReactNativeFeatureFlags.js) which contains runtime feature toggles for the Javascript layer.
* [ReactFeatureFlags.java](https://github.com/facebook/react-native/blob/main/packages/react-native/ReactAndroid/src/main/java/com/facebook/react/config/ReactFeatureFlags.java) which contains runtime feature toggles for Android.
* [ReactNativeConfig.cpp](https://github.com/facebook/react-native/blob/main/packages/react-native/ReactCommon/react/config/ReactNativeConfig.cpp) which contains runtime feature toggles for the C++ layer.

Most of those config flags are undocumented and they're not verified for invalid configuration. Specifically, some feature flags require to be toggled on multiple layers, and it's easy to forget to toggle one of them. Invalid combinations of those flags could lead to unexpected runtime behavior.

> We propose to define `reactNativeManifest` as the **preferred entry point** for all the user facing entry points. All the tooling should be adapted to consume the information from the `reactNativeManifest` section and have sensible defaults defined by the tooling itself.

Reviewer 4: Is this only for apps? I'm guessing it doesn't make sense for a library to specify flags? But I guess there could be a library that enforces a certain feature flag?

Reviewer: It's for both apps and libraries. There could be a scenarios where a capability needs all the dependencies to expose such capability (for example the `newArch` one).

Reviewer 4: Would this proposal enforce this? Like if a library specifies a certain feature flag, and an app uses that library, how do we communicate this?

Reviewer: > Would this proposal enforce this? Like if a library specifies a certain feature flag, and an app uses that library, how do we communicate this?

That's up to each feature to define the semantic of such flags. That's also why we need a dedicated RFC for each feature. There might be features that have a hard requirement on dependencies (like the New Architecture Renderer) and other features that have fallbacks

It's outside the scope of this RFC to declare all the new sections we intend to add, and we defer to the [Future proof and extensibility](#future-proof-and-extensibility) section for the process on how to include a new key to the `reactNativeManifest` section.

## Excluded use cases

This section lists the use cases where we believe the `reactNativeManifest` section should **not** be used.

### React Native version support

> For the time being, we're not planning to add a `version` section to the `reactNativeManifest` section for libraries. The rationale is that we believe that the `react-native` version should be specified in either the `peerDependencies` or the `engines` section of the `package.json` of the project. This follows the idiomatic way to specify the version of a dependency in a NodeJS project.

Reviewer 5: The examples below have a `version` section in the `reactNativeManifest`.  Is that a typo?

2. Each RFC should define a new section in the `reactNativeManifest` with what's the intended use case.
3. Each RFC should define what are the sensible defaults and should refrain from breaking changes.
4. Each RFC should define the deprecation strategy for previous flags/features related to the same capability.

## Complete Example

Here we present a full example of how a `reactNativeManifest` section could look like for a library and an app.

For an app the section will look as follows:
    
```json
{
  "name": "my-awesome-app",
  "version": "1.2.3",
  "reactNativeManifest": {
    "version": "1.0",
    "type": "app",
    "capabilities": {
      "hermes": {
        "enabled": true
      },
      "newArch": {
        "enabled": true
      },
      "codegenConfig": {
        "name": "AppSpecs",
        "type": "all",
        "jsSrcsDir": ".",
        "android": {
          "javaPackageName": "com.facebook.fbreact.specs"
        }
      }
    }
  }
}
```

For a library the section will look as follows:

```json
{
  "name": "my-awesome-library",
  "version": "1.2.3",
  "reactNativeManifest": {

>     "version": "1.0",

Reviewer 5: Something that came up in discussion is what is the correct way to version and show semantic changes in a native module.  Currently the docs for TurboModules imply versioning your native module using the package.json's version for[iOS](https://reactnative.dev/docs/the-new-architecture/pillars-turbomodules#ios-create-the-podspec-file) (I'm not seeing anything like this for Android).  There's some work taking place to surface incompatibility between native modules built (dev mode) into the native app and their JS counterparts bundled by Metro.  

Examples would be:
- forgetting to build in a native module and giving clearer error messages; or
- changing branch which includes a significant library change and metro rebuilding your JS but the developer forgetting to rebuild your native app.

## Questions:
1. Should native extensions be using the `package.json → reactNativeManifest → version` or some combination of both the `package.json → version` and the `reactNativeManifest → version`?
2. Would we update the docs, file tickets with significant [new architecture Native Modules implementations](https://reactnative.directory/?ios=true&android=true&isMaintained=true&isPopular=true&newArchitecture=true)

Reviewer 5: See from the discussion with Nick that `reactNativeManifest → version` is a [schema version](https://github.com/react-native-community/discussions-and-proposals/pull/588#discussion_r1087184428).

Reviewer: Yup also clarified here fe3e609f94490d9ac50213f5a16f71764c675d8a

    "type": "library",
    "capabilities": {
      "hermes": {
        "enabled": true
      },
      "newArch": {
        "enabled": true
      },
      "codegenConfig": {
        "name": "LibrarySpecs",
        "type": "all",
        "jsSrcsDir": ".",
        "android": {
          "javaPackageName": "com.facebook.fbreact.specs"
        }
      }
    }
  }
}
```

### JSON Schema

> The JSON Schema will be published as soon as we find an agreement on the key names and semantics.

Reviewer 5: Possibly publishing a validation library instead?  Something similar to [npmcli/package-json](https://www.npmjs.com/package/@npmcli/package-json) which users could use to parse, validate and evaluate the contents correctly (without having to home brew solutions).

Reviewer: It's up to the various tools. The easiest would be to use https://www.npmjs.com/package/jsonschema inside the CLI and validate the `package.json` against the schema we provide.

Reviewer: I've added a clarification about it in 68c906f36cef9af41a12172bc1c944900fb64f3b

  * As of today, there is no way to enable New Architecture for an app project for both platforms
  * Moreover, the file vs environment variable leads to a scenario where you can't know if a app supports New Architecture or not by inspecting the code (as New Architecture support is known at  `pod install` time and is not codified in the codebase)

Therefore we propose to add the `newArch` section to the `reactNativeManifest.capabilities` of **both apps and libraries** with the following semantic:
* For Apps: `newArch.enabled==true` implies that the app wants to use the New Architecture.
* For Libraries: `newArch.enabled==true` implies that the library is compatible with the New Architecture.

Tools can be built on top of this information to check that a app with `newArch.enabled==true` is not accepting libraries with `newArch.enabled==false` and warning against library that don't have the key specified.

The setup would look as follows for both apps and libraries:

```json
{
  "reactNativeManifest": {
    "capabilities": {
      "newArch": {
        "enabled": true
      }
    }
  }
}
```

> This section will allows also for split configuration between platforms:

Reviewer 6: Shouldn't we specify the precedence rules? Because a setup like:

```json
{
  "reactNativeManifest": {
    "capabilities": {
      "android": {
      "capabilities": {
        "newArch": {
          "enabled": true
        }
      }
    }
  }
}
```

Is invalid. What is the definition for iOS?

Secondly: do we want to allow for Libraries to say that they supports the new architecture only for one of the two platforms or not?

```json
{
  "reactNativeManifest": {
    "android": {
      "capabilities": {
        "newArch": {
          "enabled": true
        }
      }
    },
    "ios": {
      "capabilities": {
        "newArch": {
          "enabled": true
        }
      }
    }
  }
}
```

#### `codegenConfig` support

> Similarly to New Architecture support metadata, the `codegenConfig` is a key metadata of the New Architecture build pipeline.

Reviewer 6: ```suggestion
Similarly to the New Architecture support metadata, the `codegenConfig` is a key metadata of the New Architecture build pipeline.
```

Currently the `codegenConfig` is a **top-level** key in the `package.json` of a project. 

We propose to move this key under the `reactNativeManifest.capabilites` section as follows:
```json
{
  "reactNativeManifest": {
    "capabilities": {
      "codegenConfig": {
        "name": "MyLib",
        "type": "all",
        "jsSrcsDir": ".",
        "android": {
          "javaPackageName": "com.example.mypackage"
        }
      }
    }
  }
}
```

### Hermes Support

> Similarly to New Architecture support, the current way to enable/disable the Hermes engine is toggled by using a Gradle Property `hermesEnabled` on Android and changing the `:hermes_enabled` property in the Podfile.

Reviewer 6: ```suggestion
Similarly to the New Architecture support, the current way to enable/disable the Hermes engine is toggled by using a Gradle Property `hermesEnabled` on Android and changing the `:hermes_enabled` property in the Podfile.
```

We propose to add the `hermes` section under the `reactNativeManifest.capabilities` section as follows:

```json
{
  "reactNativeManifest": {
    "capabilities": {
      "hermes": {
        "enabled": true
      }
    }
  }
}
```

> which will also allow for split configuration between platforms as follows:

Reviewer 6: Same consideration as above.

We propose to define `reactNativeManifest` as the **preferred entry point** for all the user facing entry points. All the tooling should be adapted to consume the information from the `reactNativeManifest` section and have sensible defaults defined by the tooling itself.

It's outside the scope of this RFC to declare all the new sections we intend to add, and we defer to the [Future proof and extensibility](#future-proof-and-extensibility) section for the process on how to include a new key to the `reactNativeManifest` section.

## Excluded use cases

This section lists the use cases where we believe the `reactNativeManifest` section should **not** be used.

### React Native version support

For the time being, we're not planning to add a `version` section to the `reactNativeManifest` section for libraries. The rationale is that we believe that the `react-native` version should be specified in either the `peerDependencies` or the `engines` section of the `package.json` of the project. This follows the idiomatic way to specify the version of a dependency in a NodeJS project.

### TurboModule/Fabric toggles

> At the time of writing, we prefer to don't offer a dedicated section to toggle Fabric/TurboModule capability inside `reactNativeManifest`. The rationale is that we believe that the `newArch` section is sufficient to expose the New Architecture to users.

Reviewer 6: ```suggestion
At the time of writing, we prefer not to offer a dedicated section to toggle Fabric/TurboModule capability inside `reactNativeManifest`. The rationale is that we believe that the `newArch` section is sufficient to expose the New Architecture to users.
```

> Selectively toggling Fabric/TurboModule is a more advanced feature. We believe we'll still be offering a more advanced way to feature toggle such infrastructure pieces, but at the current state, our preference is to not expose such capability in the top level `reactNativeManifest` section.

Reviewer 6: ```suggestion
Selectively toggling Fabric/TurboModule is a more advanced feature. We believe we'll still be offering a more advanced way to enable or disable those pieces of the New Architecture infrastructure, but at the current state, our preference is not to expose such capability in the top level `reactNativeManifest` section.
```

### Build Logic support

Build tools such as Gradle/CocoaPods or others should account for the `reactNativeManifest` section and use it to configure the build pipeline accordingly to this and future RFCs. Specifically they should follow this logic for a generic capability `foo`.

1. If the capability `foo` is specificed in the `reactNativeManifest.capabilities` section, use the value specified in the `reactNativeManifest.capabilities.foo` section.
   1. If the user is **also** specifying the capability `foo` in the build tool specific configuration (e.g. inside `gradle.properties` for Android) behave as follows:
      1. If the two values are **compatible**, use them without notifying the user.
      1. If the two values are **incompatible**, notify the user and use the value specified in the `reactNativeManifest.capabilities.foo` section will prevail and the value specified in the build tool specific configuration will be ignored.
   1. If the user is **not** specifying the capability `foo` in the build tool specific configuration, honor the value specified in the `reactNativeManifest.capabilities.foo` section.

> 1. If the capability `foo` is **not** specificed in the `reactNativeManifest.capabilities` section, honor the value specified in the build tool specific configuration or the default value if not specified.

Reviewer 6: Can we be a little bit more aggressive?
So, for example, we can remove all the configuration flags in Cocoapods starting from version 0.73 (for example) so that the `reactNativeManifest` is the ONLY way to configure the app.
If there is no such file, we fail the build early, instructing the user how to set up those fields.

I know, it is a breaking change, but it will make the code much easier to write and follow and will let us converge to the ideal state faster. If the error messages are clear and actionable, most of the users will self-serve themselves.

### `align-deps` support

> This RFC originated from a [conversation between Meta and Microsoft](https://github.com/microsoft/rnx-kit/issues/1863) to use `align-deps` as key feature in the New Architecture rollout support, in helping developers understanding if a library they're using is New Architecture compatible or not. Very quickly it became clear that there is currently no straightforward way to know if a library indeed has that support.

Reviewer 6: ```suggestion
This RFC originated from a [conversation between Meta and Microsoft](https://github.com/microsoft/rnx-kit/issues/1863) to use `align-deps` as key feature in the New Architecture rollout support, in helping developers understand if a library they're using is compatible with the New Architecture or not. Very quickly it became clear that there is currently no straightforward way to know if a library indeed has that support.
```

> Therefore we believe `align-deps` can benefit from this information and be extended to offer New Architecture support information.

Reviewer 6: ```suggestion
Therefore we believe `align-deps` can benefit from this information and can be extended to offer New Architecture support information.
```

> ### React Native CLI support

Reviewer 6: would this be a part of React Native core or a feature we want to defer to frameworks?

The React Native direcotry can easily consume the `reactNativeManifest` by:
1. Query the NPM registry with [their API](https://github.com/npm/registry/blob/master/docs/REGISTRY-API.md) to obtain the repository information.
2. Query the repository information to obtain the `packge.json` and read the `reactNativeManifest` section.

## Future proof and extensibility

We're versioning the schema using the `schema` key in the `reactNativeManifest` section. This allows us to evolve the schema in the future without breaking changes. We believe that the `schema` key should be a semver version string.

We believe that the `reactNativeManifest` section should be extensible as we believe new capabilities and features can be added there.
Specifically, we envision to evolve `reactNativeManifest` or `reactNativeManifest.capabilites` as follows:

1. Each addition of a new capability [requires a new RFC](https://github.com/react-native-community/discussions-and-proposals#proposals) to be approved.
2. Each RFC should define a new section in the `reactNativeManifest` with what's the intended use case.
3. Each RFC should define what are the sensible defaults and should refrain from breaking changes.

> 4. Each RFC should define the deprecation strategy for previous flags/features related to the same capability.

Reviewer 6: ```suggestion
4. Each RFC should define the deprecation strategy for previous flags/features related to the same capability, if needed.
```

    "type": "library",
    "capabilities": {
      "hermes": {
        "enabled": true
      },
      "newArch": {
        "enabled": true
      },
      "codegenConfig": {
        "name": "LibrarySpecs",
        "type": "all",
        "jsSrcsDir": ".",
        "android": {
          "javaPackageName": "com.facebook.fbreact.specs"
        }
      }
    }
  }
}
```

### JSON Schema

The JSON Schema will be published as soon as we find an agreement on the key names and semantics.

## Alternatives

`reactNativeManifest` is, by design, fully original and new: there are other files that are used to define configurations for various aspect of react-native based projects (such as `react-native.config.js`, `app.json`, Expo's `app.config.js`, `expo-module.config.json`) but this one does **not** overlap with any of them. Explicitly, this one has the purpose of being filled only with capabilities and metadata to be read by package managers and build tools.

## Adoption strategy

If we decide to adopt this proposal, we'll have to broadcast this change to the ecosystem.
Ideally, we'll start by adding the `reactNativeManifest` section to the app template and to `create-react-native-library`.

> We foresee a phase in which only a part of the userbase has this section configured and tool developers need to account for the absence of this section.

Reviewer 6: ```suggestion
We foresee a phase in which only a part of the userbase has this section configured, and tool and developers need to account for the absence of this section.
```
